doc/configuration.rst
.. _configuration:

Configuration
#############
``pinnwand`` is configured in three ways, one is by arguments and the other is
through a configuration file and the last is by environment variables.

The options available are dependent on the command you're running. You can
always pass the ``--configuration-path`` argument to ``pinnwand``.

Here is a quick example::

  pinnwand --configuration-path /tmp/foo.toml http --port 9000

The ``http`` subcommand takes a separate argument ``--port`` to override
the default listening port (8000).

Any value in this file can be overriden by setting its equivalent in the environment or in a `.env` file.

To do that, all environment variables need to be in the form of `PINNWAND_{{option}}`.
For example, if we wanted to override the `database_uri` option, we will define our value like the following::

   PINNWAND_DATABASE_URI="hello"

.. note::
    * Any variable in the actual environment will take precedence over its equivalent in the `.env` file.
    * It's best to place the `.env` at the root directory of the project, to avoid including it into VCS.

File
****
The configuration file has a bunch more properties to configure ``pinnwand``
with. Here's an example file::

  # Example configuration file for `pinnwand`. Shows what you can adjust. More
  # information can be found at `pinnwand`'s documentation:
  # https://pinnwand.readthedocs.io/en/latest/ or you can file an issue at our
  # GitHub: https://github.com/supakeen/pinnwand
  
  # Database URI to connect to see: https://docs.sqlalchemy.org/en/13/core/engines.html#database-urls
  # if you want to use databases other than sqlite be sure to install the
  # appropriate packages and then format this url to correspond to them.
  database_uri = "sqlite:///:memory:"
  
  # Maximum paste size you want to allow.
  paste_size = 262144  # 256kB in bytes
  
  # Preferred lexers. This list of lexers will appear on top of the dropdown
  # on the website allowing you to preselect commonly used lexers. Note that the
  # names here have to be the identifiers used by pygments, not the human names.
  # The keys returned by /api/v1/lexer are an exhaustive list of supported lexers.
  # If empty no preferred lexers are shown.
  preferred_lexers = []

  # The lexer selected by default when creating a new paste.
  # Similar to preferred_lexers, only supports lexer identifiers used by pygments.
  default_selected_lexer = "text"

  # Logo path, used to render your logo. If left out the default logo will be
  # used. This file must be a png file.
  # logo_path = "/path/to/a/file.png"
  
  # The page path is used to find the pages listed in the page_list
  page_path = "/tmp"
  
  # This is the whitelist of files that should exist in the `page_path`
  # configuration directive. `pinnwand` will look for `$file.rst` in the
  # `page_path` directory and serve it at `/$file`.
  page_list = ["about", "removal", "expiry"]
  
  # The footer in raw HTML, shown at the bottom of the page and handy to link to
  # your previously configured pages.
  footer = 'View <a href="//github.com/supakeen/pinnwand" target="_BLANK">source code</a>, the <a href="/removal">removal</a> or <a href="/expiry">expiry</a> stories, or read the <a href="/about">about</a> page.'
  
  # HTML for the 'help text' that can be shown above the paste area, if left
  # empty no help text will be shown.
  paste_help = "<p>Welcome to pinnwand, this site is a pastebin. It allows you to share code with others. If you write code in the text area below and press the paste button you will be given a link you can share with others so they can view your code as well.</p><p>People with the link can view your pasted code, only you can remove your paste and it expires automatically. Note that anyone could guess the URI to your paste so don't rely on it being private.</p>"

  # Email used for file reporting. If the value is not None then a href with a mailto link will be added to every paste page thus allowing the users to report pastes that may need removal.
  report_email = "maintainer@example.com"

  # Expiries are given by a name and their duration in seconds, if you want to do
  # 'forever' set a really large number...
  expiries.1hour = 3600
  expiries.1day = 86400
  expiries.1week = 604800
  
  # These are application level ratelimits, if you use proxies for your pinnwand
  # instance you should set limits there as well. These limits describe a token
  # bucket and are per-IP. 
  #
  # The capacity is how many tokens there are available, the consumption is how
  # many tokens are used by an action and the refill is how many tokens are added
  # per second. So the read bucket below allows a burst of a 100 reads then another
  # 2 per second.
  ratelimit.read.capacity = 100
  ratelimit.read.consume = 1
  ratelimit.read.refill = 2
  
  ratelimit.create.capacity = 2
  ratelimit.create.consume = 2
  ratelimit.create.refill = 1
  
  ratelimit.delete.capacity = 2
  ratelimit.delete.consume = 2
  ratelimit.delete.refill = 1
  
  # pinnwand uses a naive anti-spam measure where a regex is ran over the text
  # that is pasted. It then checks how large a percentage of the incoming bytes
  # consist of links. If that percentage is larger than the number below the
  # paste is denied. Set to a 100 to disable.
  spamscore = 50

  # pinnwand has a background job that deletes all expired pastes in order to not have an overloaded database.
  # The period every which the task is executed is expressed in milliseconds
  reaping_periodicity = 1_800_000

Options
*******

database_uri
============
A URI as accepted by sqlalchemy for the database to use.

Default: ``sqlite:///:memory```

paste_size
==========
Maximum size of a formatted paste. This includes the HTML as generated by
pygments. The size should be supplied in bytes.

Default: ``262144`` (256 kB).

preferred_lexers
================
The lexers that are shown on the homepage above all other lexers. This allows
you to customize your homepage to the most-used lexers for your users.

Leaving this list empty will not show any preferred lexers. The lexer names
in this list must be supported by pygments.

Default: ``[]``.

default_selected_lexer
======================
The lexer that is selected by default when creating a new paste.

Default: ``text``

logo_path
=========
Path to a custom logo file. Needs to be readable by the user ``pinnwand`` runs
as. Leave out of the configuration file if you want to use the default logo.

Default: ``unset``.

page_path
=========
A filesystem path where pages listed in ``page_path`` are looked up in. If
unset the default ``pinnwand`` path will be used.

Default: ``unset``.

page_list
=========
List of static text pages. If set these pages will be looked up in the
``page_path`` variable. These files should exist in ``page_path`` with a
``.rst`` suffix.

Default: ``["about", "removal", "expiry"]``

footer
======
HTML to render in the footer.

Default: ``bunch of html``

paste_help
==========
HTML to render above the new paste page to help users on how to use your
instance.

Default: ``bunch of html``

report_email
============

An email address that allows users to report a paste that may need removal or
edition.

Default: ``None``

expiries
========
Several expiries exist, these are shown in the drop down by name and are used
for reaping pastes. They are denoted in seconds and the keys are free to choose.


expiries.1hour
^^^^^^^^^^^^^^
Default: `3600`

expiries.1day
^^^^^^^^^^^^^
Default: `86400`

expiries.1week
^^^^^^^^^^^^^^
Default: `604800`

ratelimit
=========
These are application level ratelimits, if you use proxies for your pinnwand
instance you should set limits there as well. These limits describe a token
bucket and are per-IP. 

The capacity is how many tokens there are available, the consumption is how
many tokens are used by an action and the refill is how many tokens are added
per second. So the read bucket below allows a burst of a 100 reads then another
2 per second.

ratelimit.read.capacity
^^^^^^^^^^^^^^^^^^^^^^^
Default: `100`

ratelimit.read.consume
^^^^^^^^^^^^^^^^^^^^^^
Default: `1`

ratelimit.read.refill
^^^^^^^^^^^^^^^^^^^^^
Default: `2`

ratelimit.create.capacity
^^^^^^^^^^^^^^^^^^^^^^^^^
Default: `2`

ratelimit.create.consume
^^^^^^^^^^^^^^^^^^^^^^^^
Default: `2`

ratelimit.create.refill
^^^^^^^^^^^^^^^^^^^^^^^
Default: `1`

ratelimit.delete.capacity
^^^^^^^^^^^^^^^^^^^^^^^^^
Default: `2`

ratelimit.delete.consume
^^^^^^^^^^^^^^^^^^^^^^^^
Default: `2`

ratelimit.delete.refill
^^^^^^^^^^^^^^^^^^^^^^^
Default: `1`

spamscore
=========
pinnwand uses a naive anti-spam measure where a regex is ran over the text
that is pasted. It then checks how large a percentage of the incoming bytes
consist of links. If that percentage is larger than the number below the
paste is denied. Set to a 100 to disable.

Default: `50`

reaping_periodicity
=========
pinnwand has a background job that deletes all expired pastes in order to not have an overloaded database.
The period every which the job is executed is expressed in milliseconds

Default: `1_800_000`


CHANGELOG.rst
Changelog
#########

``pinnwand`` is Python pastebin software that tried to keep it simple but got
a little more complex.

v1.6.0 (unreleased)
*******************
* Added a browser test suite by nekhvoya_.
* Minify the container images by williamjmorenor_.
* Fix project metadata in `pyproject.toml` by NeilHanlon_.
* A new `show` route in the v1 API to view a paste.
* Logging has been reworked by shtlrs_.
* Deprecated `log.warn` has been fixed by wookie184_.
* Better user-facing errors by wookie184_.
* Added new buttons on the view paste page nekhvoya_.
* Fix custom logo handlers to actually show the logo.
* Various configuration and log reworks and refactoring by shtlrs_.

v1.5.0 (20230727)
*****************
* Change the tab size in the HTML to be 4.
* List lexer endpoint in configuration by ChrisLovering_.
* Reorder Containerfile to optimise caching by ChrisLovering_.
* Logo is now limited on height, allowing for banner-style logos by ChrisLovering_.
* Make environment variable parsing better by accepting `SyntaxError` by ChrisLovering_.
* Allow selecting the default lexer in configuration by ChrisLovering_.
* Ctrl+S now submits pastes by ChrisLovering_.
* `pdm` is now used as the build backend and package solution.
* Containers are now published by CI to GHCR_ by ChrisLovering_.
* Add line range(s) highlight support.
* Add file upload (drag-and-drop or file picker) support.

v1.4.0 (20221127)
*****************
Focus on prettier pages, ease of development, and more defensive posturing. The
`reap` subcommand is now deprecated and instead ran in-process.

A big change is the fact that you can now configure ``pinnwand`` through
environment variables, this makes running in containerized systems a fair bit
easier.

For _packagers_, if you previously shipped timer units for `reap` you can remove
them from your packages in this new version.

* A `Containerfile` is now provided by default. This allows developers to build
  a container.
* If only a single file is pasted the code area for this single file is now
  larger.
* Update major versions of some dependencies.
* Replace the `toml` dependency with the maintained `tomli`.
* Provide example systemd service and timer file for reaping expired pastes.
* The CSS is now generated from SASS, this means that you need `sassc`
  installed if you want to change/rebuild assets.
* The `reap` subcommand is deprecated and will be removed in `1.6`, it now
  runs when the `http` subcommand is running every 1800 seconds instead. The
  example files for `reap` have been removed as well.
* ``pinnwand`` can now be configured through environment variables, these
  start with `PINNWAND_` and are all upper case, then the prefix is removed
  and the key lowercased. This *overwrites* any previously set value.
* `isort` was added to `pre-commit` checks.
* The package layout has been changed to a src-layout, if you are running tests
  with a `pytest < 7` you will need the additional `pytest-srcpaths` plugin.
* Configuration examples have moved to the `etc/` subdirectory.

v1.3.2 (20220711)
*****************
Changing some dependencies to make packaging easier.

v1.3.1 (20220220)
*****************
Bumping some major dependencies to make packaging easier.

* Update major versions of many dependencies.

v1.3.0 (20210522)
*****************
More quality of life and code quality changes.

* Provide archive download of a full paste (#92)
* Add a ``resyntax`` command to rerun lexer over all pastes (#70)
* Paste expiry options now come from the configuration file (#53)
* Convert tabs/enters to indentation, contributed by millefalcon_ (#90)
* Add line highlighting, contributed by erlliam_ (#39)
* Implement naive defensive measures (ratelimiting) (#98)

v1.2.3 (20210109)
*****************
Fixes to packaging and build setup.

* Use poetry-core as build-backend (#101)
* Remove dependency on ``poetry-dynamic-versioning`` as it breaks build.

v1.2.2 (20200829)
*****************
Some longer standing bugs get squatted in pragmatic ways.

* Ensure minimum body width to prevent button falling off (#79)
* Select initial lexer for additional files (#97)

v1.2.1 (20200806)
*****************
Minor updates to the underlying build system for easier packaging by
distros.

* Set build system correctly (#93)
* Update version number in pyproject.toml (#93)

v1.2.0 (20200806)
*****************
New features all around, minor bugfixes, code quality improvements.

* Add language autodetection, contributed by mweinelt_. (#83)
* Provide a hex view for pastes. (#86)
* Add copy to clipboard button. (#87)
* Command line option (-v) to change log level. (#88)
* Make the outline color of the focused form elements be in-line with the
  general highlight color.
* Sum up filesizes and check against paste size. This change now makes the
  paste size limit the total size, not a per-file limit! Adjust your
  configuration accordingly. (#89)
* Add a report link for files that may be problematic, this link will be
  added only if the ``report_email`` field is set to anything than None in the
  configuration file, contributed by Bruce1347_ (#2)
* Expanded testcase coverage for website from 69% to 84% by adding and fixing
  broken testcases.

v1.1.3 (20200620)
*****************
An older bug that occurs rarely resurfaced. This time a bunch of code has been
written to eradicate the problem.

* Race condition in slug_create (#34)
* Fix the millibyte notation.

v1.1.2 (20200608)
*****************
More bugfixes to some things that were either introduced in 1.1.1 or were
lower priority.

* Update our dependencies.
* Use the /static URLs directly for logo/favicon (#85)

v1.1.1 (20200602)
*****************
*The traditional bugfix release for the previous release. No real bugs here
but something to prevent CSS changes from not being loaded.*

* Prevent browsers from aggressively caching (#74)

v1.1.0 (20200524)
*****************
The 1.1.0 release is focused on new features to improve ease of use.


src/pinnwand/configuration.py
import ast
import os
from typing import Optional, TYPE_CHECKING
from dotenv import load_dotenv

if TYPE_CHECKING:  # lie to mypy, see https://github.com/python/mypy/issues/1153
    import tomllib as toml
else:
    try:
        import tomllib as toml
    except ImportError:
        import tomli as toml


class Configuration:
    """A class that holds all pinnwand's configuration."""

    def __init__(self):
        self._database_uri = "sqlite:///:memory:"
        self._paste_size = 256 * 1024  # in bytes
        self._footer = 'View <a href="//github.com/supakeen/pinnwand" target="_BLANK">source code</a>, the <a href="/removal">removal</a> or <a href="/expiry">expiry</a> stories, or read the <a href="/about">about</a> page.'
        self._paste_help = "<p>Welcome to pinnwand, this site is a pastebin. It allows you to share code with others. If you write code in the text area below and press the paste button you will be given a link you can share with others so they can view your code as well.</p><p>People with the link can view your pasted code, only you can remove your paste and it expires automatically. Note that anyone could guess the URI to your paste so don't rely on it being private.</p>"
        self._page_path = None
        self._page_list = ["about", "removal", "expiry"]
        self._default_selected_lexer = "text"
        self._preferred_lexers = []  # type: ignore
        self._logo_path = None
        self._report_email = None
        self._expiries = {"1day": 86400, "1week": 604800}
        self._reaping_periodicity = 1_800_800
        self._ratelimit = {
            "read": {
                "capacity": 100,
                "consume": 1,
                "refill": 2,
            },
            "create": {
                "capacity": 2,
                "consume": 2,
                "refill": 1,
            },
            "delete": {
                "capacity": 2,
                "consume": 2,
                "refill": 1,
            },
        }
        self._spamscore = 50

    # Define getters for each configuration parameter
    @property
    def database_uri(self):
        return self._database_uri

    @property
    def paste_size(self):
        return self._paste_size

    @property
    def footer(self):
        return self._footer

    @property
    def paste_help(self):
        return self._paste_help

    @property
    def page_path(self):
        return self._page_path

    @property
    def page_list(self):
        return self._page_list

    @property
    def default_selected_lexer(self):
        return self._default_selected_lexer

    @property
    def preferred_lexers(self):
        return self._preferred_lexers

    @property
    def logo_path(self):
        return self._logo_path

    @property
    def report_email(self):
        return self._report_email

    @property
    def expiries(self):
        return self._expiries

    @property
    def reaping_periodicity(self):
        return self._reaping_periodicity

    @property
    def ratelimit(self):
        return self._ratelimit

    @property
    def spamscore(self):
        return self._spamscore

    def load_config_file(self, path: Optional[str] = None) -> None:
        """Load configuration settings from a toml file."""

        with open(path, "rb") as file:
            loaded_configuration = toml.load(file)

            for key, value in loaded_configuration.items():
                setattr(self, f"_{key}", value)

    def load_environment(self, dotenv_file_path: Optional[str] = None):
        """Load configuration from the environment, if any."""
        prefix = "PINNWAND_"
        load_dotenv(dotenv_path=dotenv_file_path, override=False)

        for key, value in os.environ.items():
            if not key.startswith(prefix):
                continue

            key = key[len(prefix) :]
            key = key.lower()
            try:
                value = ast.literal_eval(value)
                setattr(self, f"_{key}", value)
            except (ValueError, SyntaxError):
                # When `ast.literal_eval` can't parse the value into another
                # type we take it at string value
                pass


class ConfigurationProvider:
    """A class responsible for providing an instance of the configuration on demand."""

    _config: Configuration = None

    @classmethod
    def get_config(cls, load_env=False):
        """Return the loaded configuration."""
        if not cls._config:
            cls._config = Configuration()
            if load_env:
                cls._config.load_environment()

        return cls._config
